src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx

```diff
@@ -92,51 +139,286 @@ export class FixedTimeTab extends React.Component<FixedTimeTabProps, FixedTimeTa
   }
 
   validate(): boolean {
-    const dateRange = this.createDateRange();
-    if (!dateRange) return false;
+    let dateRanges = List();
+    const { timeRanges, numTimeRanges } = this.state;
+    if (numTimeRanges === 1) {
+      const { start, end } = this.state;
+      let dateRange = this.createDateRange(start, end);
+      if (!dateRange) return false;
+      else {
+        dateRanges = dateRanges.push(dateRange);
+      }
+    } else {
+      for (let i = 0; i < numTimeRanges; i++) {
+        let date = timeRanges.get(i);
+        let dateRange = this.createDateRange(date[0], date[1]);
+        if (!dateRange) return false;
+        else {
+          dateRanges = dateRanges.push(dateRange);
+        }
+      }
+    }
     if (!this.validateTimeShift()) return false;
     const { essence: { filter, timeShift } } = this.props;
     const newTimeShift = this.constructTimeShift();
-    const newFilter = this.constructFixedFilter(dateRange);
+    const newFilter = this.constructFixedFilter(dateRanges);
     return !filter.equals(newFilter) || !timeShift.equals(newTimeShift);
   }
 
   onOkClick = () => {
+    const { timeRanges, start, end, numTimeRanges } = this.state;
     if (!this.validate()) return;
     const { clicker, onClose } = this.props;
-    clicker.changeFilter(this.constructFixedFilter(this.createDateRange()));
-    clicker.changeComparisonShift(this.constructTimeShift());
-    onClose();
+    if (timeRanges.size === 0) {
```

Comment:
Here you're checking `size` of `timeRanges` and you have `numTimeRanges` in scope. Is there a difference between them? I don't like this inferring from values that are derived from true state. In state we should have just `List` of ranges.